src/Database/Statement/BufferedStatement.php

```diff
@@ -14,54 +14,199 @@
  */
 namespace Cake\Database\Statement;
 
+use Cake\Database\StatementInterface;
+use Cake\Database\TypeConverterTrait;
+use Iterator;
+
 /**
  * A statement decorator that implements buffered results.
  *
  * This statement decorator will save fetched results in memory, allowing
  * the iterator to be rewound and reused.
  */
-class BufferedStatement extends StatementDecorator
+class BufferedStatement implements Iterator, StatementInterface
 {
+    use TypeConverterTrait;
 
     /**
-     * Records count
+     * If true, all rows were fetched
      *
-     * @var int
+     * @var bool
+     */
+    protected $_allFetched = false;
+
+    /**
+     * The decorated statement
+     *
+     * @var \Cake\Database\StatementInterface|\PDOStatement
      */
-    protected $_count = 0;
+    protected $statement;
 
     /**
-     * Array of results
+     * The driver for the statement
+     *
+     * @var \Cake\Database\DriverInterface
+     */
+    protected $driver;
+
+    /**
+     * The in-memory cache containing results from previous iterators
      *
      * @var array
      */
-    protected $_records = [];
+    protected $buffer = [];
 
     /**
-     * If true, all rows were fetched
+     * Whether or not this statement has already been executed
      *
      * @var bool
      */
-    protected $_allFetched = false;
+    protected $_hasExecuted = false;
 
     /**
-     * Current record pointer
+     * The current iterator index.
      *
      * @var int
      */
-    protected $_counter = 0;
+    protected $index = 0;
 
     /**
-     * Execute the statement and return the results.
+     * Constructor
      *
-     * @param array|null $params list of values to be bound to query
-     * @return bool true on success, false otherwise
+     * @param \Cake\Database\StatementInterface $statement Statement implementation such as PDOStatement
+     * @param \Cake\Database\Driver $driver Driver instance
+     */
+    public function __construct($statement, $driver)
+    {
+        $this->statement = $statement;
+        $this->driver = $driver;
+    }
+
+    /**
+     * Magic getter to return $queryString as read-only.
+     *
+     * @param string $property internal property to get
+     * @return mixed
+     */
+    public function __get($property)
+    {
+        if ($property === 'queryString') {
+            return $this->statement->queryString;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function bindValue($column, $value, $type = 'string')
+    {
+        $this->statement->bindValue($column, $value, $type);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function closeCursor()
+    {
+        $this->statement->closeCursor();
```

Comment:
Should we set the _allFetched variable here?

Reply:
I don't know.. I can add a test for iterating a closed statement.